vita3k/host/src/pup.cpp

```diff
@@ -0,0 +1,473 @@
+        fileout.write(result.c_str(), pad_len);
+
+        at += pad_len;
+
+        filein.seekg(segment_infos[idx].offset);
+        unsigned char *dat = new unsigned char[segment_infos[idx].size];
+        filein.read((char *)dat, segment_infos[idx].size);
+
+        unsigned char *decrypted_data = new unsigned char[segment_infos[idx].size];
+        if (segment_infos[idx].plaintext == SecureBool::NO) {
+            aes_context aes_ctx;
+            aes_setkey_enc(&aes_ctx, (unsigned char *)scesegs[i].key.c_str(), 128);
+            size_t ctr_nc_off = 0;
+            unsigned char ctr_stream_block[0x10];
+            aes_crypt_ctr(&aes_ctx, segment_infos[idx].size, &ctr_nc_off, (unsigned char *)scesegs[i].iv.c_str(), ctr_stream_block, dat, decrypted_data);
+        }
+
+        if (segment_infos[idx].compressed == SecureBool::YES) {
+            mz_stream stream;
+            stream.zalloc = Z_NULL;
+            stream.zfree = Z_NULL;
+            stream.opaque = Z_NULL;
+            stream.avail_in = 0;
+            stream.next_in = Z_NULL;
+            if (mz_inflateInit(&stream) != MZ_OK) {
+                LOG_ERROR("inflateInit failed while decompressing");
+                return;
+            }
+
+            std::string compressed_data((char *)decrypted_data, segment_infos[idx].size);
+            stream.next_in = (Bytef *)compressed_data.data();
+            stream.avail_in = compressed_data.size();
+
+            int ret;
+            char outbuffer[524288];
+            std::string decompressed_data;
+
+            do {
+                stream.next_out = reinterpret_cast<Bytef *>(outbuffer);
+                stream.avail_out = sizeof(outbuffer);
+
+                ret = mz_inflate(&stream, 0);
+
+                if (decompressed_data.size() < stream.total_out) {
+                    decompressed_data.append(outbuffer, stream.total_out - decompressed_data.size());
+                }
+
+            } while (ret == MZ_OK);
+
+            mz_inflateEnd(&stream);
+
+            if (ret != MZ_STREAM_END) {
+                LOG_ERROR("Exception during zlib decompression: ({}) {}", ret, stream.msg);
+                return;
+            }
+            fileout.write(decompressed_data.c_str(), decompressed_data.length());
+            at += decompressed_data.length();
+        } else {
+            fileout.write((char *)decrypted_data, segment_infos[idx].size);
+            at += segment_infos[idx].size;
+        }
+        delete[] dat;
+        delete[] decrypted_data;
+    }
+    filein.close();
+    fileout.close();
+}
+
+std::string make_filename(unsigned char *hdr, int64_t filetype) {
+    uint32_t magic;
+    uint32_t version;
+    uint32_t flags;
+    uint32_t moffs;
+    uint64_t metaoffs;
+    memcpy(&magic, &hdr[0], 4);
+    memcpy(&version, &hdr[4], 4);
+    memcpy(&flags, &hdr[8], 4);
+    memcpy(&moffs, &hdr[12], 4);
+    memcpy(&metaoffs, &hdr[16], 8);
+
+    if (magic == 0x454353 && version == 3 && flags == 0x30040) {
```

Comment:
Make those constant number to have a name

Perhaps SCE_MAGIC = 0x454353 or something?